lib/Backend.hs

```diff
@@ -0,0 +1,43 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module Backend
+  ( BackendEffect (..), readSecret, writeSecret, listSecrets, deleteSecret
+  , Backend (..)
+  , BackendPacked (..)
+  )
+where
+
+import qualified Data.Text           as T
+import qualified Entry               as E
+import qualified Toml
+import qualified Data.HashMap.Strict as HS
+
+import Error                         (CofferError)
+import Polysemy.Error                (Error)
+import Toml                          (TomlCodec)
+
+import Polysemy
+
+-- @TODO - rename Secret to Entry?
+data BackendEffect m a where
+  WriteSecret  :: E.Entry -> BackendEffect m Int
+  ReadSecret   :: [T.Text] -> Maybe Int -> BackendEffect m (Maybe E.Entry)
+  ListSecrets  :: [T.Text] -> BackendEffect m (Maybe [T.Text])
+  DeleteSecret :: [T.Text] -> BackendEffect m ()
```

Comment:
Can you please add haddocks to each operation? In particular, it would be useful to document operational details and invariants that weren't/can't be encoded at the type-level.

For example, we should document that:
1. `WriteSecret` overwrites any entry that might already exist at that path
2. `WriteSecret` does _not_ overwrite directories. If a directory with that path already exists, you'll end up with an entry _and_ a directory sharing the same path.
3. `ListSecret` returns path segments: if the segment is suffixed by `/` then that indicates a directory; otherwise it's an entry.
4. Once all entries are deleted from a directory, then the directory disappears (i.e. `ListSecrets` will no longer list that directory).

There may be more important details I'm missing.

Reply:
Have written some haddoc